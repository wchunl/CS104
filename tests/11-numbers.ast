ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1) function
|   |   TOK_TYPE_ID "" (6.22.1) typeid
|   |   |   TOK_VOID "void" (6.22.1) void
|   |   |   TOK_IDENT "fail" (6.22.6) variable
|   |   TOK_PARAM "(" (6.22.11) param
|   |   |   TOK_TYPE_ID "" (6.22.12) typeid
|   |   |   |   TOK_STRING "string" (6.22.12) string
|   |   |   |   TOK_IDENT "expr" (6.22.19) string variable (6.22.12)
|   |   |   TOK_TYPE_ID "" (6.22.25) typeid
|   |   |   |   TOK_STRING "string" (6.22.25) string
|   |   |   |   TOK_IDENT "file" (6.22.32) string variable (6.22.25)
|   |   |   TOK_TYPE_ID "" (6.22.38) typeid
|   |   |   |   TOK_INT "int" (6.22.38) int
|   |   |   |   TOK_IDENT "line" (6.22.42) int variable (6.22.38)
|   TOK_FUNCTION "" (6.24.1) function
|   |   TOK_TYPE_ID "" (6.24.1) typeid
|   |   |   TOK_VOID "void" (6.24.1) void
|   |   |   TOK_IDENT "putchr" (6.24.6) variable
|   |   TOK_PARAM "(" (6.24.13) param
|   |   |   TOK_TYPE_ID "" (6.24.14) typeid
|   |   |   |   TOK_INT "int" (6.24.14) int
|   |   |   |   TOK_IDENT "chr" (6.24.18) int variable (6.24.14)
|   TOK_FUNCTION "" (6.25.1) function
|   |   TOK_TYPE_ID "" (6.25.1) typeid
|   |   |   TOK_VOID "void" (6.25.1) void
|   |   |   TOK_IDENT "putint" (6.25.6) variable
|   |   TOK_PARAM "(" (6.25.13) param
|   |   |   TOK_TYPE_ID "" (6.25.14) typeid
|   |   |   |   TOK_INT "int" (6.25.14) int
|   |   |   |   TOK_IDENT "num" (6.25.18) int variable (6.25.14)
|   TOK_FUNCTION "" (6.26.1) function
|   |   TOK_TYPE_ID "" (6.26.1) typeid
|   |   |   TOK_VOID "void" (6.26.1) void
|   |   |   TOK_IDENT "putstr" (6.26.6) variable
|   |   TOK_PARAM "(" (6.26.13) param
|   |   |   TOK_TYPE_ID "" (6.26.14) typeid
|   |   |   |   TOK_STRING "string" (6.26.14) string
|   |   |   |   TOK_IDENT "str" (6.26.21) string variable (6.26.14)
|   TOK_FUNCTION "" (6.28.1) function
|   |   TOK_TYPE_ID "" (6.28.1) typeid
|   |   |   TOK_INT "int" (6.28.1) int
|   |   |   TOK_IDENT "getchr" (6.28.5) variable
|   |   TOK_PARAM "(" (6.28.11) param
|   TOK_FUNCTION "" (6.29.1) function
|   |   TOK_TYPE_ID "" (6.29.1) typeid
|   |   |   TOK_STRING "string" (6.29.1) string
|   |   |   TOK_IDENT "getstr" (6.29.8) variable
|   |   TOK_PARAM "(" (6.29.14) param
|   TOK_FUNCTION "" (6.30.1) function
|   |   TOK_TYPE_ID "" (6.30.1) typeid
|   |   |   TOK_STRING "string" (6.30.1) string
|   |   |   TOK_IDENT "getln" (6.30.8) variable
|   |   TOK_PARAM "(" (6.30.13) param
|   TOK_FUNCTION "" (7.5.1) function
|   |   TOK_TYPE_ID "" (7.5.1) typeid
|   |   |   TOK_INT "int" (7.5.1) int
|   |   |   TOK_IDENT "main" (7.5.5) variable
|   |   TOK_PARAM "(" (7.5.9) param
|   |   TOK_BLOCK "{" (7.5.12)
|   |   |   TOK_VARDECL "=" (7.6.15) variable lval
|   |   |   |   TOK_INT "int" (7.6.4) int
|   |   |   |   TOK_IDENT "number" (7.6.8) int variable (7.6.15)
|   |   |   |   TOK_INTCON "1" (7.6.17) int const
|   |   |   TOK_VARDECL "=" (7.7.13) variable lval
|   |   |   |   TOK_INT "int" (7.7.4) int
|   |   |   |   TOK_IDENT "done" (7.7.8) int variable (7.7.13)
|   |   |   |   TOK_INTCON "0" (7.7.15) int const
|   |   |   TOK_WHILE "while" (7.8.4)
|   |   |   |   TOK_NOT "not" (7.8.11)
|   |   |   |   |   TOK_IDENT "done" (7.8.15) int variable (7.7.13)
|   |   |   |   TOK_BLOCK "{" (7.8.21)
|   |   |   |   |   TOK_CALL "(" (7.9.14)
|   |   |   |   |   |   TOK_IDENT "putint" (7.9.7) variable
|   |   |   |   |   |   TOK_IDENT "number" (7.9.15) int variable (7.6.15)
|   |   |   |   |   TOK_CALL "(" (7.10.14)
|   |   |   |   |   |   TOK_IDENT "putchr" (7.10.7) variable
|   |   |   |   |   |   TOK_CHARCON "'\n'" (7.10.15) int const
|   |   |   |   |   TOK_IF "if" (7.11.7)
|   |   |   |   |   |   TOK_LE "<=" (7.11.18) vreg
|   |   |   |   |   |   |   TOK_IDENT "number" (7.11.11) int variable (7.6.15)
|   |   |   |   |   |   |   TOK_INTCON "0" (7.11.21) int const
|   |   |   |   |   |   '=' "=" (7.11.29) int vreg
|   |   |   |   |   |   |   TOK_IDENT "done" (7.11.24) int variable (7.7.13)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.11.31) int const
|   |   |   |   |   '=' "=" (7.12.14) int vreg
|   |   |   |   |   |   TOK_IDENT "number" (7.12.7) int variable (7.6.15)
|   |   |   |   |   |   '+' "+" (7.12.23) int vreg
|   |   |   |   |   |   |   TOK_IDENT "number" (7.12.16) int variable (7.6.15)
|   |   |   |   |   |   |   TOK_IDENT "number" (7.12.25) int variable (7.6.15)
|   |   |   TOK_RETURN "return" (7.14.4)
|   |   |   |   TOK_INTCON "0" (7.14.11) int const
