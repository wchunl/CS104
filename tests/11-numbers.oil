fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
main:     .function int
          .local int number
          .local int done
          number = 1
          done = 0
.wh0:     $t0:i = not done
          goto .od0 if not $t0:i
.do0:     putint(number)
          putchr('\n')
.if0:     $t1:i = number <= 0
          goto .el0 if not $t1:i
.th0:     done = 1
.fi0:     $t2:i = number + number
          number = $t2:i
.od0:     return 0
          return 
          .end
