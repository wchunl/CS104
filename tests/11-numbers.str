string_set[   0]:    4706206694200543916 0x1983c08->","
                   16579916199511856337 0x1985098->"getln"
string_set[   1]:    2297668033614959926 0x1985aa8->"0"
string_set[   3]:   11868120863286193964 0x19852b8->"main"
                    9611428142580219387 0x197f608->"fail"
string_set[   4]:      21606623943142546 0x1985508->"{"
                    4017852018407011111 0x1983cf8->"file"
string_set[   6]:   13686721134187162920 0x1986318->"}"
string_set[   8]:    1370730858159036685 0x19860c8->"+"
                    9386170080689500517 0x1983ad8->"expr"
string_set[   9]:   10319632053171904911 0x19858b8->"while"
string_set[  12]:   15816506932648033945 0x1983968->"("
string_set[  14]:    9317282556992397945 0x1985ef8->"if"
string_set[  17]:   15430489877420048373 0x1986228->"<="
                   18413295556186398326 0x19840f8->"putint"
string_set[  18]:    7211130878751977441 0x1984228->"putchr"
                   12679825168387489523 0x1984708->"num"
string_set[  23]:   11166701219533395393 0x1984e58->"getstr"
                   12868445110721718657 0x1983d18->")"
string_set[  28]:   14635384904416252381 0x1985b48->"return"
string_set[  29]:   12291392672316406109 0x1983ea8->"line"
string_set[  31]:    6192799629375248351 0x1986028->"'\n'"
                    6142509188972423790 0x197f5e8->""
                   13348923631108115385 0x19844e8->"putstr"
string_set[  32]:   11597697714117577063 0x1985778->"="
string_set[  33]:    3182526723088623345 0x1984038->"number"
                    1583576014053361350 0x1984888->"str"
string_set[  35]:   14849783028996024063 0x1985c58->"not"
string_set[  38]:   15578157668962758274 0x1985998->"done"
                   12597418120160561823 0x1983b48->"int"
string_set[  39]:    7296690267450279874 0x1984368->"chr"
string_set[  40]:   11911796782241914524 0x197f218->"void"
string_set[  42]:    5344662657899890615 0x1983ec8->";"
string_set[  44]:   13497578857207247412 0x1983a08->"string"
string_set[  45]:   10159970873491820195 0x1985818->"1"
string_set[  46]:    9201800968376702500 0x1984c38->"getchr"
load_factor = 0.766
bucket_count = 47
max_bucket_size = 3
