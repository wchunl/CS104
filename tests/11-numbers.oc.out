; 0.  1: # 1 "11-numbers.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "11-numbers.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "11-numbers.oc" 2
; 7.  4: 
; 7.  5: int main() {
; 7.  6:    int number = 1;
; 7.  7:    int done = 0;
; 7.  8:    while (not done) {
; 7.  9:       putint (number);
; 7. 10:       putchr ('\n');
; 7. 11:       if (number <= 0) done = 1;
; 7. 12:       number = number + number;
; 7. 13:    }
; 7. 14:    return 0;
; 7. 15: }
hello